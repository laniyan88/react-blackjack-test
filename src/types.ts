export enum CardSuit {
  Clubs = "clubs",
  Diamonds = "diamonds",
  Hearts = "hearts",
  Spades = "spades",
}

export enum CardRank {
  Ace = "ace",
  Two = "2",
  Three = "3",
  Four = "4",
  Five = "5",
  Six = "6",
  Seven = "7",
  Eight = "8",
  Nine = "9",
  Ten = "10",
  Jack = "jack",
  Queen = "queen",
  King = "king",
}

export type Card = { suit: CardSuit; rank: CardRank };
export type Deck = Card[];
export type Hand = Card[];

export type BlackjackState = {
  deck: Deck;
  player: Hand;
  dealer: Hand;
  gameOver: boolean;
  message: string;
  dealerStatus: "" | "win";
  playerStatus: "" | "win";
  draw: boolean;
};
