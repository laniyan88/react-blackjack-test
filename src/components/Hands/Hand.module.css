.wrapper {
  display: flex;
  flex-direction: row-reverse;
  align-items: center;
  justify-content: flex-start;
  margin: 0.75rem 0;
  padding-left: 1rem;
  gap: 0.5rem;
}

.score {
  font-size: 1rem;
  background-color: #004b2d;
  padding: 1rem;
  border-radius: 5rem;
  width: 3rem;
  height: 3rem;
  display: flex;
  justify-content: center;
  align-items: center;
  border: inset #ffd700 0.2rem;
}

.score > h3 {
  color: white;
}

.win {
  background-color: #ffd700;
}

@media (max-width: 30em) {
  .wrapper {
    padding-left: 0.5rem;
    gap: 0.4rem;
  }
}
