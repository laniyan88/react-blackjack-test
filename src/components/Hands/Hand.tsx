import React from "react";
import { Hand as HandType } from "../../types";
import { getDealerStartValue, getHandValue } from "../../utils/scoring";
import { CardImage } from "../CardImage/CardImage";
import styles from "./Hand.module.css";

export const Hand: React.FC<{
  cards: HandType;
  highlightWin?: boolean;
  label: string;
  showHole?: boolean;
}> = ({ cards, highlightWin, label, showHole = true }) => {
  const score = showHole ? getHandValue(cards): getDealerStartValue(cards);
  const STAGGER_MS = 60;

  return (
    <div className={styles['wrapper']} aria-label={`${label} hand`}>
      <div className={`${styles['score']} ${highlightWin ? styles['win'] : ""}`} aria-label={`${label} score`}>
        <h3>{score}</h3>
      </div>
      {cards.map((c, i) => (
        <CardImage key={i} {...c} hidden={i > 0 && !showHole} appearDelay={i * STAGGER_MS}/>
      ))}
    </div>
  );
};
