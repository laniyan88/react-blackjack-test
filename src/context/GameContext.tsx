import React, { createContext, useContext } from "react";
import { useBlackjackController } from "../hooks/useBlackjack";

export type GameController = ReturnType<typeof useBlackjackController>;

const GameContext = createContext<GameController | null>(null);

export const GameProvider: React.FC<React.PropsWithChildren<{}>> = ({ children }) => {
  const controller = useBlackjackController({ topUp: 100, keyboard: true });
  return <GameContext.Provider value={controller}>{children}</GameContext.Provider>;
};

export const useGame = (): GameController => {
  const ctx = useContext(GameContext);
  if (!ctx) {
    throw new Error("useGame must be used within a GameProvider");
  }
  return ctx;
};
