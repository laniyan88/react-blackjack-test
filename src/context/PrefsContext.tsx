import React, { createContext, useContext, useEffect, useMemo, useState } from "react";
import { LS_KEYS } from "../utils/constant";
import { safeStorage } from "../utils/safeStorage";

type Prefs = {
  cardBack: string;
  setCardBack: (name: string) => void;
  cardBackUrl: string;
  availableBacks: string[];
};

const PrefsContext = createContext<Prefs | null>(null);

const BACKS = [
  "back.png",
  "back-black.png",
  "back-blue.png",
  "back-fuchsia.png",
  "back-gray.png",
  "back-green.png",
  "back-lime.png",
  "back-maroon.png",
  "back-navy.png",
  "back-olive.png",
  "back-purple.png",
  "back-red.png",
  "back-silver.png",
  "back-teal.png",
  "back-yellow.png",
  "back-aqua.png",
  "back-0062ff.png",
];

export const PrefsProvider: React.FC<React.PropsWithChildren<{}>> = ({ children }) => {
  const [cardBack, setCardBack] = useState<string>(() => {
    const saved = safeStorage.getItem(LS_KEYS.cardBack);
    return saved && BACKS.includes(saved) ? saved : "back-blue.png";
  });

  useEffect(() => {
    try {
      safeStorage.setItem(LS_KEYS.cardBack, cardBack);
    } catch {}
  }, [cardBack]);

  const cardBackUrl = useMemo(
    () => `${process.env.PUBLIC_URL}/SVG-cards/png/1x/${cardBack}`,
    [cardBack],
  );

  const value: Prefs = useMemo(
    () => ({ cardBack, setCardBack, cardBackUrl, availableBacks: BACKS }),
    [cardBack, cardBackUrl],
  );

  return <PrefsContext.Provider value={value}>{children}</PrefsContext.Provider>;
};

export const usePrefs = (): Prefs => {
  const ctx = useContext(PrefsContext);
  if (!ctx) throw new Error("usePrefs must be used within PrefsProvider");
  return ctx;
};
