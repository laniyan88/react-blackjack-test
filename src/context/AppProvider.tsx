import React from "react";
import { PrefsProvider } from "./PrefsContext";
import { GameProvider } from "./GameContext";

export function AppProviders({ children }: { children: React.ReactNode }) {
    return (
        <PrefsProvider>
            <GameProvider>
                {children}
            </GameProvider>
        </PrefsProvider>
    )
}