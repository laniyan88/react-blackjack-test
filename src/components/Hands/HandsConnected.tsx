import React from "react";
import { Hand } from "./Hand";
import { WIN } from "../../utils/constant";
import { useGame } from "../../context/GameContext";

export const HandsConnected: React.FC = () => {
  const { state, dealerHoleRevealed } = useGame();
  if (!state) return null;

  return (
    <>
      <Hand
        label="Dealer"
        cards={state.dealer}
        showHole={dealerHoleRevealed}
        highlightWin={state.dealerStatus === WIN}
      />
      <Hand
        label="Player"
        cards={state.player}
        showHole={true}
        highlightWin={state.playerStatus === WIN}
      />
    </>
  );
};
