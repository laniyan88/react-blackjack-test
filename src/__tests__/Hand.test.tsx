import React from "react";
import { render, screen } from "@testing-library/react";
import { CardRank, CardSuit } from "../types";
import { PrefsProvider } from "../context/PrefsContext";
import { Hand } from "../components/Hands/Hand";

const C = (rank: CardRank, suit: CardSuit = CardSuit.Clubs) => ({ rank, suit });

describe("Hand", () => {
  it("shows score and cards; snapshot", () => {
    const cards = [C(CardRank.Ace), C(CardRank.Nine)];
    const { asFragment } = render(
      <PrefsProvider>
        <Hand label="Player" cards={cards} showHole={true} highlightWin={false} />
      </PrefsProvider>
    );

    expect(screen.getByLabelText(/player score/i)).toHaveTextContent("20");
    expect(screen.getAllByRole("img").length).toBeGreaterThanOrEqual(2);
    expect(asFragment()).toMatchSnapshot();
  });
});
