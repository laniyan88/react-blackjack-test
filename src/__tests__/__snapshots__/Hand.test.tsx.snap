// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hand shows score and cards; snapshot 1`] = `
<DocumentFragment>
  <div
    aria-label="Player hand"
    class="wrapper"
  >
    <div
      aria-label="Player score"
      class="score "
    >
      <h3>
        20
      </h3>
    </div>
    <div
      aria-hidden="false"
      class="flip enter"
    >
      <div
        class="flipInner "
        style="animation-delay: 0ms;"
      >
        <img
          alt="ace of clubs"
          class="face front"
          draggable="false"
          src="/SVG-cards/png/1x/club_1.png"
        />
        <img
          alt="Face-down card"
          class="face back"
          draggable="false"
          src="/SVG-cards/png/1x/back-blue.png"
        />
      </div>
    </div>
    <div
      aria-hidden="false"
      class="flip enter"
    >
      <div
        class="flipInner "
        style="animation-delay: 60ms;"
      >
        <img
          alt="9 of clubs"
          class="face front"
          draggable="false"
          src="/SVG-cards/png/1x/club_9.png"
        />
        <img
          alt="Face-down card"
          class="face back"
          draggable="false"
          src="/SVG-cards/png/1x/back-blue.png"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
